Baidu CTC wrapper

joaovictortr opened the pull request:
Fix and adapt an existing wrapper for [Baidu CTC](https://github.com/mcf06/theano_ctc), then integrate it into tensor.nnet.

If the [warp-ctc](https://github.com/baidu-research/warp-ctc) library is not on the compiler's default library path, one should set the ```ctc.root``` variable to point to a local copy of the aforementioned library.

Before executing the CTC wrapepr, it is necessary to have the library libwarpctc.so on the LD path, so one also has to change the LD_LIBRARY_PATH environment variable to include the directory that contains the library.

For example:
```sh
export LD_LIBRARY_PATH=$CTC_LIB/build:$LD_LIBRARY_PATH
THEANO_FLAGS=`floatX=float32,ctc.root=/home/user/warp-ctc' python theano/tensor/nnet/tests/test_ctc.py
``` 

Tasks:

- [x] Port CPU implementation (done using a COp)
- [x] Add tests and verification of results for CPU implementation (test done, more test cases required)
- [x] CPU to GPU optimization
- [x] Fix and port GPUOp to new backend
- [x] Add tests and verification of results for GPU implementation
- [x] Use OpenMP in CPU wrapper
- [x] Use CUDA stream in GPU wrapper

fix #3871

joaovictortr commented:
@nouiz @notoraptor, I've opened this PR, so it is possible to keep track of progress in the CTC wrapper.

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,76 @@
+import os
+
+class ConnectionistTemporalClassification(gof.COp):
+    __props__ = ()
```
`"computeGradient"` should be a prop.

Also, to keep syntax coherent, I suggest to use `compute_gradient` instead of `computeGradient` for the name of that class member !

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,76 @@
+        if self.computeGradient:
+            self.gradients = T.ftensor3(name="ctc_grad")
+
+        if "CTC_LIB" in os.environ:
```
Insted of using environment variables, we generally add a new theano config flag that is used only when it is specified. This makes configuration easier in user side:
 - if the config flag is not specified, g++ will look for headers and libraries in default's OS or Python folders.
 - else Theano will get convenient folders from config flag. This allows user to change library folder quickly if needed.

Here I suggest to add config flag `ctc.root` equivalent to your current `CTC_LIB` environment variable. Check `theano/configdefaults.py` to see how we add a new config variable: https://github.com/Theano/Theano/blob/master/theano/configdefaults.py

As an example, check how we use flags `dnn.include_path` and `dnn.library_path` in new back-end DNN module: `theano/gpuarray/dnn.py`: https://github.com/Theano/Theano/blob/master/theano/gpuarray/dnn.py#L255 .

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,76 @@
+        self.default_output = 0 
+
+        return gof.Apply(self, inputs=[t_activations, t_labels, t_input_lengths],
+                         outputs=[self.costs, self.gradients])
```
`self.gradients` may not exists if `self.computeGradient` is false !

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,209 @@
+    else if ( PyArray_NDIM( *out_costs ) != 1 ||
+              PyArray_DIMS( *out_costs )[0] != cost_size )  // matrix has the wrong size
+    {
+        Py_XDECREF( *out_costs ); 
```
Note that `Py_XDECREF(pointer)` works also with NULL `pointer` (it does nothing in such case). So, this 
`if(A) {allocate} else if(B) {xdecref then allocate}` 
may be compacted to 
`if (A or B) {xdecref then allocate}`.

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,209 @@
+    
+    costs = (npy_float32 *) PyArray_DATA( *out_costs );
+
+    if ( NULL == (*out_gradients) )
```
Same remark here (this `if -- else if` may take less lines of code !).

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,209 @@
+        return 1;    
+    }
+
+    ctc_cpu_workspace = malloc( cpu_workspace_size );
```
Check if ctc_cpu_workspace is not NULL after malloc.

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,209 @@
+        return 1;         
+    }
+
+    if ( NULL != activations_copy )
```
If an error occurs above, the function will return without deallocating `activation_copy` if needed.

In general, make sure that, at every point where an error occurs, any allocated memory is deallocated before the function returns!

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,76 @@
+        # output_grads[0].shape = [batchSize]  (one cost per sequence)
+        # So, reshape output_grads to [1, batchSize, 1] for broadcasting
+        output_grad = output_grads[0].reshape( (1, -1, 1) )
+        return [output_grad * self.gradients,
```
Same here (`self.gradients` may not exists if we have not `self.computeGradient`).

notoraptor commented:
NB: I have forgotten: for CTC integration, I suggest you also add a theano config flag like `ctc.enabled`. This will help skip CTC tests on machines when CTC is not enabled. Look for `magma.enabled` flag into `theano/configdefaults.py` and `theano/gpuarray/tests/test_linalg.py` as example.

Tell us if you have any problem with theano config flags !

joaovictortr commented:
I've added ctc.enabled and ctc.root configuration parameters, I'll also use the ctc.enabled flag to skip the test if CTC is not enabled. computeGrad was changed to compute_grad, as well.

There is also a dummy OpenMP Op to pull in OpenMP headers and compilation options, without having to rely on multiple inheritance or changing from COp to OpenMPOp.

> self.gradients may not exists if we have not self.computeGradient
self.compute_grad (formerly self.computeGradient) is only false if the user does not request the gradient, through the use of a local optimization.

A problem remains though, when the function is compiled, it cannot find the library (libwarpctc.so).  I've checked the paths and everything is correct -- ctc.enabled is True and ctc.root is set to the local copy of warp-ctc. Do you have any idea on what the problem might be?

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,242 @@
+                      ctc_msg );
+        *status = 1;
+    }
+    *status = 0;
```
`*status` will always contain 0, as this line is always executed even if we enter in the `if` block above. Also, this function should return an int, but it currently returns nothing.

I think you want to return the status, isn't it ? So, you could remove the `*status` arg, and then return 1 if retcode is not CTC_STATUS_SUCCESS, and 0 otherwise. Or keep `*status` arg and set this function as a void function.

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,242 @@
+                                 PyArrayObject *  in_labels,
+                                 PyArrayObject *  in_input_lengths,
+                                 PyArrayObject ** out_costs,
+                                 PyArrayObject ** out_gradients)
```
So, this function will always try to return an out_gradients, but in Python code, there is an optimization that set the output to be only out_costs if gradient is not used. I think this should make the execution crash (as out_gradients arg will receive nothing here, not even NULL).

You may find these issues when you will run more exhaustive tests. Don't forget to test both cases (when gradient is used and when it is not used)!

joaovictortr replied:
I've added both cases in the refactored test, I'll test as soon as I can solve the issue with libwarpctc.so

abergeron replied:
This can be fixed with the `_cop_num_outputs` attribute on the class.  Set it to 2 and the second output argument will be set to NULL if not provided on the node.

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,242 @@
+    {
+        PyErr_Format( PyExc_MemoryError,
+            "Could not allocate storage for input lengths" );
+        return 1;        
```
`activations_copy` may have been allocated at this stage, too !

To help manage memory allocation/deallocation and avoid memory leaks, what I can suggest is to create a struct that will contain all pointers that you may want to allocate and deallocate. For example, a struct called `CPUCtctCostContext` with an init function (`CPUCtctCostContext_init(CPUCtctCostContext*)` that set any pointer member to NULL), and a clear function (`CPUCtctCostContext_clear(CPUCtctCostContext*)`, which free any pointer member that is not null). You can put this struct and its functions in `#section support_code` section.

Then, in `APPLY_SPECIFIC(ctc_cost_cpu)`, you create the struct on stack (`CPUCtctCostContext context;`), you initialize it (`CPUCtctCostContext_init(&context)`), you allocate members when you need (e.g. `context.activations_copy = ...;`) and then, everywhere an error occurs (like here), you just call `CPUCtctCostContext_clear(&context)` just before returning.

notoraptor commented:
**NB**: Also, OpenMPOp does additional works (see OpenMPOp.prepare_node()). It is indeed a little tricky, but you should better inherit from both OpenMPOp and COp.

I suggest to do that later, when CPU op will be ok and safe (after tests are implemented and any bug fixed). The most important is to have working Ops first, and then we could add parallelization when possible and if useful.

joaovictortr commented:
Hello @notoraptor, I've changed the type of the configuration parameter to StrParam, however the problem still persists. I'll keep looking into it.

I've also ran flake8 through the Python code and fixed the errors I've encountered.

The test file should now be in the format of the other Theano tests, however more test cases are still necessary.

joaovictortr commented:
@notoraptor I've also added the ctc_context_t type and its auxiliary functions to manage the memory in the CTC wrapper context

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,292 @@
+        "Failed to obtain CTC workspace size!" );
+
+    if ( ctc_error )  // Exception is set by ctc_check_result, return error here
+        return 1;
```
We should call `ctc_context_destroy()` before returning here, as ctc context may contain allocated memory.

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,292 @@
+        context->options ), "Failed to compute CTC loss function!" );
+
+    if ( ctc_error )  // Exception is set by ctc_check_result, return error here
+        return 1;
```
We should call `ctc_context_destroy()` before returning here, as ctc context may contain allocated memory.

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,292 @@
+                                 PyArrayObject ** out_costs,
+                                 PyArrayObject ** out_gradients)
+{
+    ctc_context_t * context = (ctc_context_t *)malloc( sizeof( ctc_context_t ) );
```
It may be more convenient to declare context on stack instead of allocating it (on heap). Allocation (malloc) implies deallocation (free), so that you will be forced to call `free(context)` after `ctc_context_destroy(context)` everywhere you want to return.

Instead, you can replace this line with:
```C
// Object on stack.
ctc_context_t context_object;
/*
Pointer to object on stack
(with this, you won't need to replace `context` by `&context` in rest of code).
*/
ctc_context_t* context = &context_object;
```

Then, you won't need to call `free(context)` anywhere, as `context_object` itself will be automatically destroyed by C (because it is on stack).

You still have to call `ctc_context_destroy()` to free internal context pointers!

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,292 @@
+        return 1;
+
+    ctc_context_destroy( context );
+    free( context );
```
Then this `free()` can be removed!

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,292 @@
+ * Wrapper version with gradient computation disabled.
+ *
+ * We assume that function overloading (C++) is available to distinguish between
+ * two versions of the ctc_cost_cpu function.
```
Let's avoid using C++ overloading ! Give an unique name to each function !

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,292 @@
+    int * label_lengths;
+    PyArrayObject * activations_copy;
+    PyArrayObject * costs;
+    PyArrayObject * gradients;
```
In fact, we can ignore these 3 PyArrayObjects:
 - You won't need activations_copy anymore with the new strategy I suggest.
 - We generally don't try to deallocate an output that we have allocated, because this output will be retrieved by Python code of Theano, and will be cleared here. So you don't need to worry about freeing costs and gradients if an error occurs.

So you can just remove them from the context and forget them.

joaovictortr commented:
Hello @notoraptor, I've modified the code to include your suggestions and corrections:
* Moved allocation of CTC context from heap back to stack
* Added missing context deallocations on errors
* Use a separate function for when gradient computation is disabled (ctc_cost_cpu_no_grad)
* Remove the PyArrayObject pointers from the context
* Ensure activations array is C-contiguous by using cpu_contiguous Op
* Remove the activations copy inside the wrapper

joaovictortr commented:
Hello @notoraptor , there still seems to be a problem with the value returned from the test. I've checked with the pure C implementation (out of Theano), and there is discrepancy in the first value of the cost. In the current test case, it returns a value around 3, when it really should be 1.6. I've checked the code to be sure, and didn't found the cause yet. Do you have any ideas on what might be the issue here?

notoraptor commented:
Hello, @joaovictortr . From the example you use ( https://github.com/baidu-research/warp-ctc/blob/master/torch_binding/TUTORIAL.md ), the code for Torch should be:
```
th>acts = torch.Tensor({{0,0,0,0,0},{1,2,3,4,5},{-5,-4,-3,-2,-1},
                        {0,0,0,0,0},{6,7,8,9,10},{-10,-9,-8,-7,-6},
                        {0,0,0,0,0},{11,12,13,14,15},{-15,-14,-13,-12,-11}}):cuda()
th>labels = {{1}, {3,3}, {2,3}}
th>sizes = {1,3,3}
th>gpu_ctc(acts, grads, labels, sizes)
```
So, first label sequence should be {1}. But in Theano code you use {1, -1}, and it seems the problem comes from here. NumPy requires to have an array with all rows having same lengths, so you must find the right value to add after `1`, and it seems `-1` is not the right value to use.

notoraptor commented:
PS: My bad, the problem is not about `-1`, as labels are re-wrapped in C code. Still investigating.

notoraptor commented:
@joaovictortr I remark that CTC Torch example uses a 2-D matrix for activations, while your code uses a 3-D tensor to store activations.

```lua
-- Here it's a 2D matrix with shape (9, 5).
th>acts = torch.Tensor({{0,0,0,0,0},{1,2,3,4,5},{-5,-4,-3,-2,-1},
                        {0,0,0,0,0},{6,7,8,9,10},{-10,-9,-8,-7,-6},
                        {0,0,0,0,0},{11,12,13,14,15},{-15,-14,-13,-12,-11}}):cuda()
```

```python
# Here it's a 3-D tensor with shape (3, 3, 5).
        activations = np.asarray([[[0, 0, 0, 0, 0], [1, 2, 3, 4, 5], [-5, -4, -3, -2, -1]],
                                  [[0, 0, 0, 0, 0], [6, 7, 8, 9, 10], [-10, -9, -8, -7, -6]],
                                  [[0, 0, 0, 0, 0], [11, 12, 13, 14, 15], [-15, -14, -13, -12, -11]]],
                                 dtype=np.float32)
```

Maybe error comes from here, as CTC expects a 2-D matrix ? This could involve to rewrite some parts of your C code.

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,108 @@
+        t_activations = cpu_contiguous(t_activations)
+
+        t_labels = T.as_tensor_variable(labels)
+        t_input_lengths = T.cast(activations.shape[0], dtype="int32") * \
```
In fact, you should use `input_lengths` if it is not None, and compute it (as you do here) only if it is None.
```python
        if input_lengths is None:
            t_input_lengths = T.cast(activations.shape[0], dtype="int32") * \
                T.ones_like(activations[0, :, 0], dtype=np.int32)
        else:
            t_input_lengths = T.as_tensor_variable(input_lengths)
```

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,108 @@
+
+        # Return only the cost. Gradient will be returned by grad()
+        self.default_output = 0
+
```
Check that t_activations, t_labels and t_input_lengths have expected dtypes (float32, int32 and int32).

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,265 @@
+}
+
+void create_contiguous_input_lengths( PyArrayObject * input_lengths_arr,
+    int ** input_lengths )
```
For input_lengths, use `npy_int32` instead of `int`, as `int` may not be 32-bits on all computers. Update this everywhere in the code.

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,265 @@
+        int label_length = 0;
+        for( int col_idx = 0; col_idx < cols; ++col_idx )
+        {
+            int label = *( (int *) PyArray_GETPTR2( label_matrix, row_idx, col_idx ) );
```
label_matrix pointer should be casted to npy_int32 instead of int, as label_matrix should be an int32 tensor (checked in make_node()). It could then be dereferenced and copied to an int without any loss, so the rest of code for this function should be still ok.

```C
int label = *( (npy_int32 *) PyArray_GETPTR2( label_matrix, row_idx, col_idx ) );
```

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,265 @@
+
+    for( int elem_idx = 0; elem_idx < num_elements; ++elem_idx )
+    {
+        (*input_lengths)[elem_idx] = *( (int *) PyArray_GETPTR1( input_lengths_arr, elem_idx ) );
```
input_lenghts_arr pointer should be casted to npy_int32 instead of int, as input_lenghts_arr should be an int32 tensor (checked in make_node()). It could then be dereferenced and copied to an int without any loss, so the rest of code for this function should be still ok.
```C
(*input_lengths)[elem_idx] = *( (npy_int32 *) PyArray_GETPTR1( input_lengths_arr, elem_idx ) );
```

joaovictortr commented:
Hello @notoraptor, I've made the corrections you've suggested.

> I found the real error, see comments below. In fact, you have just forgotten to use input_lengths when it is given by the user.

Indeed that was the problem, thanks a lot for your quick feedback, mate!

I just have to check the result of the gradients (it is differing from the C implementation), and add the remaining test cases.

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,264 @@
+    int ctc_error;
+
+    ctc_error = ctc_check_result( get_workspace_size( context->label_lengths,
+        context->input_lengths, alphabet_size, minibatch_size, context->options,
```
Well, now `context->label_lengths` and `context->input_lengths` are `npy_int` arrays, ie. `npy_int32`(4 bytes). and `get_workspace_size()` expects them to be `int` arrays ( https://github.com/baidu-research/warp-ctc/blob/master/include/ctc.h#L123 ), but an `int` can be 2, 4 or even 8 bytes depending on hardware, operating system and compiler.

To avoid any future problems, I suggest than `context->label_lengths`, `context->input_lengths` and `context->flat_labels` be all `int`, so that they will be always passed as array of current `int` to CTC functions:

```C
typedef struct ctc_context {
    struct ctcOptions options;
    void * workspace;
    int * input_lengths; // int
    int * flat_labels; // int
    int * label_lengths; // int
} ctc_context_t;
```

Then, in functions `create_contiguous_input_lengths()` and `create_flat_labels()`, when you extract values from PyArrayObject, you cast PyArrayObject data to its real pointer and then you convert the value you got to an int.

```C
void create_contiguous_input_lengths( PyArrayObject * input_lengths_arr, /* int */ int ** input_lengths )
{
    npy_int num_elements = PyArray_DIMS( input_lengths_arr )[0];

    // Allocate an array of int, not npy_int
    *input_lengths = (int *) malloc( num_elements * sizeof(npy_int) );

    if ( NULL == (*input_lengths) )
        return;

    for( npy_int elem_idx = 0; elem_idx < num_elements; ++elem_idx )
    {
        // This could stay as-is, as value will be implicitely casted to int.
        (*input_lengths)[elem_idx] = *( (npy_int *) PyArray_GETPTR1( input_lengths_arr, elem_idx ) );
    }
}

void create_flat_labels( PyArrayObject * label_matrix, /*int*/ int ** flat_labels,
    /*int*/ int ** label_lengths )
{
    
    // ...
    
    // Allocate an array of int, not npy_int
    *flat_labels = (int *) malloc( rows * cols * sizeof(npy_int) );
    //...

    // Allocate an array of int, not npy_int
    *label_lengths = (int *) malloc( rows * sizeof(npy_int) );

    // ...
            
            // This could stay as-is, as value will be implicitely casted to int.
            npy_int label = *( (npy_int *) PyArray_GETPTR2( label_matrix, row_idx, col_idx ) );
            
            // ...

}
```

So, we keep the types expected by CTC API, but we do conversions when we extract values from PyArrayObjects.

The last thing to do is to make sure that we know the types of PyArrayObjects (see my remark in `make_node()`).

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,107 @@
+
+        t_labels = T.as_tensor_variable(labels)
+        t_input_lengths = T.as_tensor_variable(input_lengths)
+
```
We still don't know if inputs have expected types, so check it at this point.
```python
if t_activations.type.dtype != 'float32':
    raise ... # we expect activations to be float32 (=> npy_float32)
if t_labels.type.dtype != 'int32'
    raise ... # we expect labels to be int32 (=> npy_int)
if t_input_lengths.type.dtype != 'int32':
    raise ... # we expect input lengths to be int32 (=> npy_int)
```

joaovictortr commented:
Hello @notoraptor , I have found the issue in the gradient results and incorporated the changes you've request.
Changes:

* Types of numpy arrays are now checked in CTC.
* Compatibility with Baidu CTC's API is maintained by using int pointers instead o npy_int.
* Another test case was added to test the CPU implementation

Do you have any more suggestions? Also, should I add more (or what kinds of) tests?

notoraptor commented on theano/gpuarray/tests/test_ctc.py:
```diff
@@ -0,0 +1,58 @@
+        test = theano.function([], [t_cost])
+
+        cost, grad = train()
+        test_cost, = test()
```
It seems `test_cost` is unused, so you could probably remove that line and also the definition of theano `test`function above (line 32) .

notoraptor commented on theano/gpuarray/tests/test_ctc.py:
```diff
@@ -0,0 +1,58 @@
+        #utt.assert_allclose(expected_grads, grad)
+        #utt.assert_allclose(expected_costs, cost)
+
+    def simple_test(self):
```
Rename this method to `test_ctc(self)`.

joaovictortr commented:
@notoraptor , I've added the verify_grad test, but is still under work, and also incorporated your suggested modifications.

Currently, I cannot get the GPU wrapper to compile, because it fails to pull in the ctc.h header. I'm looking into it, but did not get very far yet. Do you have any ideas on what the issue might be?

notoraptor commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,124 @@
+
+    def c_headers(self):
+        headers = ['ctc.h']
+        headers = headers + super(CGpuKernelBase, self).c_headers()
```
Remove this line and try again. It seems parent class `GpuKernelBase` will always try to get headers from the child class.

joaovictortr commented:
Hello @notoraptor , I'm almost completing the GPU wrapper, but here are two remaining issues:

* Fix the return type of gradients in the grad function. When using the gradients passed to the C wrapper, Theano returns ```MissingInputError```.
* How to properly add the cuda include path (in order to use cudaMalloc and cudaFree). Using ```config.cuda.root``` does not seem to work with string paths.

Also, I'm using npy_float32 and npy_int for float32 and int types inside the C wrapper. Is that the proper approach?

notoraptor commented:
Hello, @joaovictortr ! I am checking your code on my computer, and `verify_grad()` is currently failing. That's quite strange as the other tests (test_ctc and test_torch_case) succeed.

How have you computed the expected gradients used in test_ctc and test_torch_case ?

notoraptor commented on theano/gpuarray/ctc_wrapper.c:
```diff
@@ -0,0 +1,254 @@
+
+    ctc_error = ctc_check_result( compute_ctc_loss( activations, gradients,
+        context->flat_labels, context->label_lengths, context->input_lengths,
+        alphabet_size, minibatch_size, costs, context->workspace,
```
Then, for workspace here, we use: `PyGpuArray_DEV_DATA(context->workspace)`.

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,113 @@
+            raise RuntimeError('Baidu CTC is not enabled and '
+                               'ConnectionistTemporalClassification Op '
+                               'can not be constructed.')
+        return [self.gradients,
```
In fact, `grad()` method in Theano is not intended to simply return the gradient of the Op's output with respect to the Op's inputs. The full documentation can be found here: http://deeplearning.net/software/theano_versions/dev/extending/op.html#gradient . I'll try to explain that in following lines:

Theano works on graphs of Ops and want to be able to compute the gradient of any output of full graph with respect to any initial input of the graph.

So, for example, let's take a simple Op that performs a function `f` so that `f(x) = y` (`x` in the Op's input, `y` is the Op's output). Then, Op's `grad()` method will receive the gradient of the graph output `C` (graph in which current Op is present) with respect to the Op's output `y`: that is, `dC/dy`. The goal of `grad()` method is to return the gradient of graph output `C` with respect to the Op's input `x`: that is, `dC/dx`. To do that, `grad()` method generally compute `dy/dx` (the real classic gradient of Op's output with respect to Op's input), and then computes `dC/dx` by multiplying `dC/dy` with `dy/dx`:
```
dC/dx = dC/dy * dy/dx
```
`dC/dy` is in `output_grads` parameters. In practice, `output_grads` contains one `dC/dy` for each Op's output `y`. You should then return `dC/dx` for each Op's input `x`.

Here, for first input (activations), `dy/dx` is indeed `self.gradients`, and `dC/dy` is `output_grads[0]` (first output is the cost), then we should "multiply" these two variables. As CTC works like a batched Op (multiple independent activations sequences), we must be careful about how we perform multiplication. I found the following code to do that:
```python
        import theano
        z = output_grads[0]
        g = theano.tensor.basic.batched_dot(z, self.gradients.dimshuffle(1, 0, 2)).dimshuffle(1, 0, 2)
```
(each cost is relative to each sequence of activations, but activations sequences are in the second dimension of activations tensor (minibatch size), so the first call to `dimshuffle()` puts that second dimension of activations to the front, so that each gradient wrt/ cost multiplies its corresponding sequence of gradients wrt activations; at the end, second call to `dimshuffle()` put the tensor back to its original dimensions order).

Then, you could return `[g, grad_undefined..., grad_undefined...]` instead of `[self.gradients, grad_undefined..., grad_undefined...]` (gradient is still undefined for other inputs, ie. labels and input lengths).

notoraptor commented on theano/tensor/nnet/tests/test_ctc.py:
```diff
@@ -0,0 +1,117 @@
+
+        cost, grad = train()
+
+        utt.assert_allclose(expected_grads, grad)
```
The grad is computed for the mean of costs, not for costs themselves. Currently, the expected gradients you give is the gradient for the cost.

To get the expected gradient for the mean, I think we could just change this line:
```python
utt.assert_allclose(expected_grads / input_length.shape[0], grad)
```
(gradient of (mean of (costs)) is mean of (gradients of (costs)), and there are `input_length.shape[0]` costs, I think)

joaovictortr commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,138 @@
+            # We assume here that the header is available at the include directory
+            # of the CTC root directory.
+            dirs.append(os.path.join(config.ctc.root, "include"))
+        dirs.append('/usr/local/cuda/include')
```
Baidu CTC library does not currently support Windows, only Mac OS X and Linux. But if the approach using only PyGpu is utilized, than the cuda path is no longer required.

joaovictortr commented:
Hello @notoraptor , you were absolutely correct, I was missing the composition of the chain rule derivative in the gradients, and that was the reason verify_grad tests were failing. Thanks for taking to write about the role of grad function.

Only one problem remains now with the GPU implementation of grad. I'm using GpuDimShuffle, and it emits an error, expecting the input of the symbolic operation GpuDimShuffle. Something like the following:

```theano.gof.fg.MissingInputError: Input 0 of the graph (indices start from 0), used to compute InplaceGpuDimShuffle{1,0,2}(<GpuArrayType<None>(float32, 3D)>), was not provided and not given```.

What would be the proper way of shuffling dimensions with GPU arrays?

notoraptor commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,164 @@
+            A 1-D tensor with the number of time steps for each sequence in
+            the minibatch.
+
+        """
```
This docstring is usually placed at class level (after `class CpuConnectionist...`).

notoraptor commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,164 @@
+                               'can not be constructed.')
+
+        context = infer_context_name(activations, labels, input_lengths)
+        assert context == self.context_name
```
Remove this assert.

notoraptor commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,164 @@
+        if self.gradients is not None:
+            out_params.append(as_gpuarray_variable(self.gradients(),
+                                                   context_name=self.context_name))
+
```
You should now create costs and gradients variables here.
```python
out_params = [GpuArrayType(...)()] # costs
if self.compute_grad:
    out_params.append(GpuArrayType(...)()) # gradients
```

notoraptor commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,164 @@
+    return GpuConnectionistTemporalClassification()(activations, labels,
+                                                    input_lengths)
+
+
```
We also need an optimization that replaces CPU CTC with GPU CTC. An example here: https://github.com/Theano/Theano/blob/master/theano/gpuarray/opt.py#L2008

notoraptor commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,164 @@
+    def c_headers(self):
+        return ['ctc.h', 'numpy_compat.h', 'gpuarray_helper.h', 'gpuarray/types.h',
+                'gpuarray_api.h', 'gpuarray/array.h', 'gpuarray/util.h']
+
```
Now, to get the GPU context in C code (last parameter of functions, `PyGpuContextObject *  ctx`), we will set the Op param.
First we define the param as Op attribute, then we override the `get_params()` method that will retrieve the context from node inputs:
```python
# gpu_context_type must be imported from gpuarray/type.py
params_type = gpu_context_type
def get_params(self, node):
    # input 0 is `activations`, which will be on GPU, so it will have a context.
    return node.inputs[0].type.context
```

notoraptor commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,164 @@
+        return theano.Apply(self, inputs=[t_activations, t_labels, t_input_lengths],
+                            outputs=out_params)
+
+    def grad(self, inputs, output_grads):
```
Then, the strategy (as we have not currently access to Op's outputs because we don't define them anymore in `__init__()`): we will override abstract function `L_op()` instead of `grad()`. `L_op()` should return the same thing as `grad()`, but the difference is that it will receive Op's outputs. This is its default implementation: https://github.com/Theano/Theano/blob/master/theano/gof/op.py#L710
```python
    def L_op(self, inputs, outputs, output_grads):
        return self.grad(inputs, output_grads)
```
(in fact, it is not documented!). Then when overriding it, we can get gradients:
```python
    def L_op(self, inputs, outputs, output_grads):
        gradients = outputs[1] # already a variable (so we just use `gradients`, not `gradients()`)
        # ....
```

notoraptor commented on theano/gpuarray/ctc_wrapper.c:
```diff
@@ -0,0 +1,284 @@
+#section kernels
```
Section to remove.

notoraptor commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,164 @@
+                               'can not be constructed.')
+        z = output_grads[0]
+        grad_shuffle = GpuDimShuffle(input_broadcastable=(False, False, False),
+                                     new_order=(1, 0, 2))(self.gradients())
```
As info, I think your errors occur because `self.gradients()` here is never included in a graph: in `__init__()`, `self.gradients` is defined as a Theano type. So, call to `self.gradients()` creates a new variable in `make_node()`, and the same call here creates another nex variable that is not used anywhere else.

But as we completely change implementation, this error should be avoided.

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,142 @@
+        self.costs = T.fvector(name="ctc_cost")
+        if self.compute_grad:
+            self.gradients = T.ftensor3(name="ctc_grad")
+
```
We should use `L_op()` for CPU op also, so that we can avoid creating variables in `__init__()`. It's better for readability and code encapsulation/reusage (as an op instance may be called many times with different inputs), and we usually create variables in `make_node()` mainly!

joaovictortr commented:
Hello @notoraptor , I've implemented all of your suggestions and remarks.

Using L_op did indeed solve the problem with shuffling the GpuArray, thanks again for pointing this essential piece of Theano's API out!

There are two remaining tasks:

* Integrate the CUDA stream into the C code
* Find out why only the test_torch_case test's gradients are wrong in the GPU wrapper

Does the GPU implementations also employ verifiy_grad tests?

notoraptor commented:
Hello, @joaovictortr ! Yes, `test_verify_grad` will work with GPU also, add this test too.

I'm lokking at the issue about `test_torch_case`, but I don't yet find the reason.

About CUDA streams, isn't it already integrated with these lines in GPU CTC code ?
```C
    context->options.loc = CTC_GPU;
    context->options.stream = 0;
```

Also, can you rebase your PR ? There is a conflict.

joaovictortr commented:
Hello @notoraptor, what I've meant with getting a stream, was in the context of multi-stream applications, it seems that it is not fully implemented in Theano yet. I'll leave the null ('0') stream for now.

Using a [C implementation](https://gist.github.com/joaovictortr/c1af19896aeabbc272bb29dc581518b8), yields the same gradient results as the ones that are being generated with the wrapper, so the underlying library is the culprit.

notoraptor commented on theano/tensor/nnet/tests/test_ctc.py:
```diff
@@ -0,0 +1,111 @@
+        expected_gradients = np.asarray(grads, dtype=np.float32)
+
+        self.run_ctc(activations, labels, activation_times, expected_costs, expected_gradients)
+
```
Also add a test to check that grad is not computed when it is not used.

notoraptor replied:
We also need a quick test to check that grad is not computed when it is not used (ie. optimization that removes gradient is applied correctly).

Just build a short theano function that computes cost only, and then check that `node.op.compute_grad` is false for the node with a CTC op.

notoraptor commented on theano/gpuarray/tests/test_ctc.py:
```diff
@@ -0,0 +1,115 @@
+        cpu_cost = np.empty(shape=cost.shape, dtype=np.float32)
+        # Transfer costs from GPU memory to host
+        cost.read(cpu_cost)
+        cost.sync()
```
You can just do `cpu_cost = np.asarray(cost)` here, without the sequence `np.empty, read, sync`. It should work.

Same comment for `gpu_grad` below.

joaovictortr commented:
Hello @notoraptor, I've included your suggestions and also the tests to check the lifter optimization. Please let me know if you have more feedback.

What else is left to merge this PR?

notoraptor commented on theano/gpuarray/opt.py:
```diff
@@ -2104,6 +2119,17 @@ def local_gpu_svd(op, context_name, inputs, outputs):
+    if not config.ctc.enabled:
+        return
+    op = GpuConnectionistTemporalClassification(compute_grad=op.compute_grad)
+    return list(op(*inputs))
```
As these ops set `default_output` to 0, only cost will be always returned when calling `op(*inputs)`, leading to only 1 output, but current node to be optimized may have 2 outputs (cost and grad) if compute_grad is true.

It's better to directly returned the result of `make_node()`, as this function will effectively return 2 outputs if compute_grad is true, and 1 output if compute_grad is false, so that we will always have exactly the same number of outputs as the current optimized node (otherwise optimization will fail).

```python
apply_node = op.make_node(*inputs)
return apply_node.outputs
```

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,148 @@
+        self.default_output = 0
+
+        return gof.Apply(self, inputs=[t_activations, t_labels, t_input_lengths],
+                         outputs=[costs, gradients])
```
Here also, build outputs list in same way as you do on GPU, ie:
```python
# pseudocode
outputs = [costs]
if compute_grad:
   outputs += [gradients]
```

(as gradients may not exist if compute_grad is false, It is important for Op consistency and reusability !).

joaovictortr commented:
Hello @notoraptor, I apologize for the several typos.

I've fixed the typos and incorporated all of your latest suggestions. Could you please take a look at it again?

notoraptor commented:
ok to test

notoraptor commented:
jenkins test this

notoraptor commented:
jenkins test this

notoraptor commented:
jenkins test this

notoraptor commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,179 @@
+from theano.tensor.opt import register_stabilize
+
+import os
+import pygpu
```
Hello, @joaovictortr , could you replace this line with:
```python
from . import pygpu
```
This will import pygpu from gpuarray init file, which should help prevent some errors with jenkins when building Theano documentation.

joaovictortr commented:
Hello @notoraptor, one of the Travis tests failed due to a timeout. Should it be executed again?

I'll continue with the Spatial Transformer Networks, as you suggested. In the meanwhile, we can discuss in the theano-dev list, over the linear algebra ops.

notoraptor commented:
I restarted the test that failed. It should pass (this happens sometimes).

notoraptor commented:
Strange recurrent failing test on TRAVIS due to a timeout. I have relaunched the test many times, still failing: https://travis-ci.org/Theano/Theano/jobs/243012604 . This test should not be related to this PR as it excludes folder `theano/tensor/nnet`. @nouiz @abergeron @lamblin .

nouiz commented:
I think this need doc inside Theano/doc to tell people about it and have link/information to install it, enable it.

nouiz commented on doc/library/config.txt:
```diff
@@ -716,6 +716,23 @@ import theano and print the config variable, as in:
+    Default: ``''``
+
+    Location of the warp-ctc folder. The folder should contain a build
+    subfolder with the shared library (libwarpctc.so), and another
```
Having a "build" subdirectory is strange. Normally, the subdirectory is named lib or lib64. Can you also support that? Why did you do differently? Normally, when a c packages is installed it is like that.

joaovictortr replied:
It can also support having lib and lib64 as include directories.

The Implementation was done this way because the build directory is the one created with the default instructions from the [warp-ctc](https://github.com/baidu-research/warp-ctc/blob/master/README.md#compilation) repository.

joaovictortr replied:
Added support for `lib`, `lib64` and `build` subdirectories within `config.ctc.root`

abergeron replied:
You should probably update the doc to reflect that.

nouiz commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,179 @@
+    """
+    __props__ = ('compute_grad',)
+
+    func_file = "./ctc_wrapper.c"
```
To help have not too many fies in the directory, we started in another PR to put the c code in the subfolder "c_code". Can you move it there?

joaovictortr replied:
No problem, does it also need to be done for the CPU version?

notoraptor replied:
Hello, @joaovictortr , do it only for GPU version for the moment.

The folder should be `theano/gpuarray/c_code`, as here: https://github.com/Theano/Theano/pull/5959/commits/5b6d3ac7bb74ef02e86d9091c80ca00e1f5c7e6d .

joaovictortr replied:
Done

nouiz commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,179 @@
+@register_canonicalize
+@register_stabilize
+@local_optimizer([GpuConnectionistTemporalClassification])
+def local_GpuConnectionistTemporalClassification_no_grad(node):
```
only lower case name in the fct name.

joaovictortr replied:
Done, I've also simplified the function name

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,177 @@
+    ----------
+    compute_grad
+        If set to True, enables the computation of gradients of the CTC loss function.
+    Returns
```
As this is a class docstring, remove the `Returns` section.

joaovictortr replied:
Removed from both CPU and GPU classes

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,177 @@
+    -------
+    1-D tensor
+        Cost of each example in the minibatch. Tensor is of shape
+        (time index, minibatch index, probabilities).
```
Is it indeed the shape of the returned 1-D tensor. It seems the cost array instead contains 1 value per batch element, isn't it ?

joaovictortr replied:
Indeed, it returns a 1D array with one cost per element in the minibatch. I've fixed the return type in the doc

notoraptor commented on doc/library/tensor/nnet/ctc.txt:
```diff
@@ -0,0 +1,28 @@
+.. moduleauthor:: `João Victor Risso <https://github.com/joaovictortr>`_
+
+.. autofunction:: theano.tensor.nnet.ctc.ctc
+.. autofunction:: theano.tensor.nnet.ctc.ConnectionistTemporalClassification
```
Here, use `autoclass::` instead, as `ConnectionistTemporalClassification` is a class.

joaovictortr replied:
I've must have overlooked it, and it is corrected now

notoraptor commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,190 @@
+    ----------
+    compute_grad
+        If set to True, enables the computation of gradients of the CTC loss function.
+    Returns
```
As this is a class docstring, remove the `Returns` section.

notoraptor commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,190 @@
+    -------
+    1-D tensor
+        Cost of each example in the minibatch. Tensor is of shape
+        (time index, minibatch index, probabilities).
```
Same remark here (see my comment in CPU code).

notoraptor commented on doc/library/tensor/nnet/ctc.txt:
```diff
@@ -0,0 +1,28 @@
+.. autofunction:: theano.tensor.nnet.ctc.ctc
+.. autofunction:: theano.tensor.nnet.ctc.ConnectionistTemporalClassification
+
+.. automodule:: theano.tensor.nnet.ctc
```
As you already organize doc in previous lines (with autofunction and autoclass), you don't need to use the automodule section here. It will just print the doc twice.

notoraptor commented:
Well, now let's wait for travis and jenkins!

joaovictortr commented:
Hello @notoraptor, there seems to be only a problem with one of the tests in Travis. It is related to a timeout in the elemwise test. Is it worth to rerun the tests again?

notoraptor commented:
@joaovictortr there are such timeout errors with other PRs also, so it does not seem to be related to that PR.

I have relaunched the test, but I think it will still fail. For the moment, we don't know why this error occurs.

botev commented:
@notoraptor Could this be soon merged into the master branch as it is of key importance for things like speech recognition.

notoraptor commented:
@botev Certainly the next week ! I just wait for a final pass of @abergeron !

joaovictortr commented:
@botev I am working on a simple nnet with Lasagne to showcase the CTC functionality, it should be available soon

abergeron commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,185 @@
+
+    def __init__(self, compute_grad=True):
+        if not compute_grad:
+            self.func_name = "APPLY_SPECIFIC(ctc_cost_gpu_no_grad)"
```
As mentioned elsewhere you should not need to do this if you use _cop_num_outputs.  It will automatically set the unavailable outputs to NULL.

abergeron commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,185 @@
+                               'GpuConnectionistTemporalClassification Op '
+                               'can not be constructed.')
+
+        context = infer_context_name(activations, labels, input_lengths)
```
Please change the name of the variable to `context_name` to avoid confusion.

Also, if you want labels and input_lengths as CPU variables, don't look at them to choose a context.

abergeron commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,185 @@
+        if not ctc_enabled:
+            raise RuntimeError('Baidu CTC is not enabled and '
+                               'GpuConnectionistTemporalClassification Op '
+                               'can not be constructed.')
```
This check is not useful because if that function is called, we already have a GpuConnectionistTemporalClassification object.

abergeron commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,185 @@
+                               'GpuConnectionistTemporalClassification Op '
+                               'can not be constructed.')
+        # Gradients computed by Op
+        gradients = outputs[1]
```
This just assumes that the gradient output is there, but it could be missing and this will crash.

abergeron commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,185 @@
+        if len(node.outputs) > 1:
+            if len(node.outputs[1].clients) == 0:   # gradient is not used
+                node.op = GpuConnectionistTemporalClassification(compute_grad=False)
+                node.outputs = node.outputs[:1]   # costs only
```
You should NEVER modify a node inplace.  Optimizations are supposed to return new outputs to replace the ones of the current node.

abergeron commented on theano/gpuarray/opt.py:
```diff
@@ -2143,6 +2158,18 @@ def local_gpu_svd(op, context_name, inputs, outputs):
+@register_opt2([ConnectionistTemporalClassification], 'ctc', 'fast_compile')
+def local_gpu_ctc(op, context_name, inputs, outputs):
+    if not config.ctc.enabled:
+        return
```
I'm not sure this check is useful, because if we get here them we already use CTC on the cpu.

abergeron commented on theano/gpuarray/c_code/ctc_wrapper.c:
```diff
@@ -0,0 +1,293 @@
+
+    // Get CUDA function pointer to obtain stream
+    CUstream (*getstream_func_ptr)(void *) = (CUstream (*)(void *)) gpuarray_get_extension( "cuda_get_stream" );
+    context->options.stream = getstream_func_ptr(gpu_context->ctx);
```
You should use the NULL stream here or add proper synchronization for all your GPU inputs/outputs in the calling code.

abergeron commented on theano/gpuarray/c_code/ctc_wrapper.c:
```diff
@@ -0,0 +1,293 @@
+    ctc_error = ctc_check_result( compute_ctc_loss( activations, gradients,
+        context->flat_labels, context->label_lengths, context->input_lengths,
+        alphabet_size, minibatch_size, costs, PyGpuArray_DEV_DATA(context->workspace),
+        context->options ), "Failed to compute CTC loss function!" );
```
This should probably be wrapped in cuda_enter()/cuda_exit() to set the proper ambiant context.  I don't know if the get_workspace_size() function above also needs this.

abergeron commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,172 @@
+
+    def __init__(self, compute_grad=True):
+        if not compute_grad:
+            self.func_name = "APPLY_SPECIFIC(ctc_cost_cpu_no_grad)"
```
_cop_num_outputs

abergeron commented on doc/library/config.txt:
```diff
@@ -728,6 +728,23 @@ import theano and print the config variable, as in:
+    If ``'True'``, use `warp-ctc <https://github.com/baidu-research/warp-ctc>`_ for CTC
+    loss computations.
+
+    If ``'False'``, disable warp-ctc.
```
This configuration value makes no sense to me.  Either you use the CTC ops in your graph and you always want them or you don't and you don't want them.  There will never be a case where you use the ops, but want to disable them.

notoraptor replied:
Currrently, it is useful to set `ctc.enabled` to `False` (done by default) to avoid jenking failures, as CTC is not yet installed on Jenkins machines.

Or, should we introduce a function like `ctc_available()` (like `dnn_available()`) to replace this configuration ?

abergeron replied:
This should be auto-detected and the tests skipped.

abergeron commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,172 @@
+            outputs += [gradients]
+
+        # Return only the cost. Gradient will be returned by grad()
+        self.default_output = 0
```
This should be a class property.

abergeron commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,185 @@
+            raise TypeError('Label lengths must use the int32 type!')
+
+        # Return only the cost. Gradient will be returned by grad()
+        self.default_output = 0
```
This should be a class property.

abergeron commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,172 @@
+            raise RuntimeError('Baidu CTC is not enabled and '
+                               'ConnectionistTemporalClassification Op '
+                               'can not be constructed.')
+        gradients = outputs[1]
```
outputs[1] can be missing in some cases.  Check for that.

abergeron commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,172 @@
+                         outputs=outputs)
+
+    def L_op(self, inputs, outputs, output_grads):
+        if not ctc_enabled:
```
Remove this check

abergeron commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,267 @@
+void ctc_context_destroy(ctc_context_t * context)
+{
+    if ( NULL != context->workspace )
+        free( context->workspace );
```
You don't need NULL checks for free().

abergeron commented on theano/gpuarray/tests/test_ctc.py:
```diff
@@ -0,0 +1,188 @@
+from .config import (mode_with_gpu, mode_without_gpu)
+
+
+class TestCTC(unittest.TestCase):
```
This copies a lot of the testing code from the CPU tests.  It should probably be refactored so that the CPU test can be reused with some parametrization.

Also, we don't really need to test direct usage since we expect people to use the CPU ops and rely on the optimizations to move to the GPU.

nouiz commented:
Ideally, we should include a ctc_available(), but this is not mandatory.

We should include in jenkins CTC before merging this PR.

On Mon, Jul 3, 2017 at 3:40 PM notoraptor <notifications@github.com> wrote:

> *@notoraptor* commented on this pull request.
> ------------------------------
>
> In doc/library/config.txt
> <https://github.com/Theano/Theano/pull/5949#discussion_r125348633>:
>
> > @@ -728,6 +728,23 @@ import theano and print the config variable, as in:
>
>      Location of the magma library.
>
> +.. attribute:: config.ctc.enabled
> +
> +    Default: ``'False'``
> +
> +    If ``'True'``, use `warp-ctc <https://github.com/baidu-research/warp-ctc>`_ for CTC
> +    loss computations.
> +
> +    If ``'False'``, disable warp-ctc.
>
> Currrently, it is useful to set ctc.enabled to False (done by default) to
> avoid jenking failures, as CTC is not yet installed on Jenkins machines.
>
> Or, should we introduce a function like ctc_available() (like
> dnn_available()) to replace this configuration ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Theano/Theano/pull/5949#discussion_r125348633>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AALC-2P-Dmawwu1F-2hpWnZmZIBAOj-wks5sKUPJgaJpZM4NdOYZ>
> .
>

notoraptor commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,179 @@
+    compute_grad
+        If set to True, enables the computation of gradients of the CTC loss function.
+    """
+    __props__ = ('compute_grad', 'default_output',)
```
`default_output` does not need to be in `__props__` list (it is more a class attribute, not an Op property).

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,167 @@
+    compute_grad
+        If set to True, enables the computation of gradients of the CTC loss function.
+    """
+    __props__ = ('compute_grad', 'default_output',)
```
`default_output` does not need to be in `__props__` list (it is more a class attribute, not an Op property).

joaovictortr commented:
Hello @notoraptor , there are still some corrections to be made, like refactoring the GPU tests. Sorry, the `default_output` prop was a mistake during the refactoring.

notoraptor commented:
Hello, @joaovictortr ! I forgot to summarize the changes we need about `ctc.enabled`:
- remove the flag `ctc.enabled`
- replace it with a function `ctc_available()` that detects at runtime if CTC is available. To implement it, you can see `_dnn_check_compile()` as example: https://github.com/Theano/Theano/blob/master/theano/gpuarray/dnn.py#L94 . The goal would be to compile a small code that call a fast CTC function (e.g. `get_warpctc_version()`), with right headers, GCC args and linked libraries.
- make sure this function does not recompile every time it is called. To do that, you could set and maintain a static function variable that will save the result at first call (e.g. `ctc_available.avail`).

notoraptor commented:
**NB**; @slefrancois we will have to work to add [Baidu's CTC library](https://github.com/baidu-research/warp-ctc) into jenkins (see @nouiz comment above: https://github.com/Theano/Theano/pull/5949#issuecomment-312722987 ) !

NB: As magma, I think it will compile on Linux only.

slefrancois commented:
Sounds good, we'll add it to the planning.

joaovictortr commented:
Hello @notoraptor , I've added the function to auto-detect warp-ctc, I've used the `ctcOptions` struct definition as test example in the compilation. Do you think that's sufficient?

notoraptor commented:
@joaovictortr it should be (as `ctcOptions` should be available only through `ctc.h` header).

Thanks ! I'll do a complete review tomorrow !

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,229 @@
+        costs = T.fvector(name="ctc_cost")
+        outputs = [costs]
+        if self.compute_grad:
+            self.gradients = T.ftensor3(name="ctc_grad")
```
Here also, just `gradients`, not `self.gradients`.

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,229 @@
+                         outputs=outputs)
+
+    def L_op(self, inputs, outputs, output_grads):
+        gradients = self.gradients
```
Using `outputs[1]` instead seems to work with your current code.

notoraptor commented on theano/gpuarray/c_code/ctc_wrapper.c:
```diff
@@ -0,0 +1,296 @@
+
+    ctc_context_init( context, gpu_context );
+
+    switch (in_activations->ga.typecode)
```
**NB**: You have not checked if in_activations is contiguous, before using it below.

joaovictortr replied:
I think it does not apply here, since we're calling ``gpu_contiguous(activations)`` in ``make_node``, prior to calling the C code.

notoraptor replied:
Indeed, it should work.

notoraptor commented:
Hi, @joaovictortr ! Just to say that this PR needs a rebase. There are also 2 remaining comments to be addressed.

joaovictortr commented:
Hi, @notoraptor ! I have made the rebase, and besides the ``default_output`` in the Ops' properties, what comment do you refer to?

notoraptor commented:
@joaovictortr about `config.ctc.enabled`. It should not be mentioned anymore in the code!

joaovictortr commented:
@notoraptor There was still one metion there, I thought I had removed everything, sorry.

notoraptor commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,162 @@
+
+        if t_input_lengths.type.dtype != 'int32':
+            raise TypeError('Label lengths must use the int32 type!')
+
```
**NB**: sorry, I forgot, but check here that inputs have expected nim (3 for activations, 2 for labels, 1 for inputs_lengths).

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,231 @@
+
+        if t_input_lengths.type.dtype != 'int32':
+            raise TypeError('Label lengths must use the int32 type!')
+
```
Here also, check ndims for inputs.

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,248 @@
+
+    if ( NULL == context->input_lengths )
+    {
+        PyErr_Format( PyExc_MemoryError,
```
Just add `ctc_context_destroy( context );` before this line. It is useless at this stage, but better for maintainability.

notoraptor commented on theano/tensor/nnet/ctc_wrapper.c:
```diff
@@ -0,0 +1,248 @@
+    npy_int rows = PyArray_DIMS( label_matrix )[0];
+    npy_int cols = PyArray_DIMS( label_matrix )[1];
+
+    *flat_labels = (int *) malloc( rows * cols * sizeof(int) );
```
Maybe `calloc` here also, as in GPU code.

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,240 @@
+            lib_found = os.path.exists(os.path.join(lib_path, "libwarpctc.so"))
+            if lib_found:
+                return True, lib_path
+    return False, None
```
Hello, @joaovictortr ! While installing CTC on jenkins, we remarked that user will necessarly have to specify `ctc.root` even if CTC is installed in default system paths (e.g. in `/usr/local` on Linux). We should be able to automatically find CTC if user installed it as a default system library.

I suggest that:
1) `_ctc_find_lib()` just returns either `lib_path` if it is found, or `None` otherwise.
2) `ctc_present()` passes the result of `_ctc_find_lib()` directly to `_ctc_check_compile()` without any checking on `ctc_lib_path`.
3) Then, in `_ctc_check_compile()`, you add `-L<libpath>` only if libpathh is not None. Otherwise, GCC will try to find `libwarpctc.so` in default systems paths.

joaovictortr replied:
Hello @notoraptor , well noted! I've included the suggestions below.

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,237 @@
+options.num_threads = 1;
+"""
+
+    params = ["-I%s" % (os.path.join(config.ctc.root, "include"))]
```
Here also, if `ctc.root` is not configured, then don't add `-I <include-path>`.

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,237 @@
+    """
+    Find the directory that contains libwarpctc.so
+    """
+    lib_found = False
```
This declaration seems useless now !

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,237 @@
+
+
+def ctc_available():
+    if config.ctc.root == '':
```
Now it should not be checked anymore.

notoraptor commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,171 @@
+
+    def c_header_dirs(self):
+        dirs = [os.path.dirname(__file__), pygpu.get_include()]
+        dirs.append(os.path.join(config.ctc.root, "include"))
```
Here also (to be appended only if ctc.root is configured).

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,237 @@
+    Find the directory that contains libwarpctc.so
+    """
+    lib_found = False
+    for lib_dir in ["build", "lib", "lib64"]:
```
In fact, you can just skip this `for` loop if ctc.root is not configured.

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,237 @@
+    Notes
+    -----
+    Using the wrapper requires that Baidu's warp-ctc library is installed and the
+    configuration variable `config.ctc.root` be properly set.
```
Update remark about ctc.root here.

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,237 @@
+    Notes
+    -----
+    Using the loss function requires that Baidu's warp-ctc library is installed and the
+    configuration variable `config.ctc.root` be properly set.
```
Update remark about ctc.root here.

notoraptor replied:
@joaovictortr Here also (update comment)!

notoraptor commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,171 @@
+
+    def c_lib_dirs(self):
+        assert ctc_available.path is not None
+        return [ctc_available.path]
```
Should now be returned only if it is not None, else return an empty line `[]`.

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,237 @@
+        self.default_output = 0
+
+    def c_lib_dirs(self):
+        assert ctc_available.path is not None
```
Should now be returned only if it is not None, else return an empty line `[]`.

notoraptor commented:
Hello @joaovictortr , please don't forget this last comment ! https://github.com/Theano/Theano/pull/5949#discussion_r127047281 .

joaovictortr commented:
Hello @notoraptor , I've removed the last pending comment, thanks for pointing it out! Please let me kwnow, if there are more changes that need to be done.

notoraptor commented on doc/library/tensor/nnet/ctc.txt:
```diff
@@ -0,0 +1,27 @@
+.. _libdoc_tensor_nnet_ctc:
+
+==============================================================
+:mod:`ctc` -- Connectionist Temporal Classification (CTC) loss
```
As @abergeron suggested, and as you do in `doc/library/gpuarray/ctc.txt`, use the full module name here (`theano.tensor.nnet.ctc`).

joaovictortr replied:
Regression fixed.

notoraptor commented on doc/library/tensor/nnet/ctc.txt:
```diff
@@ -0,0 +1,26 @@
+    Unfortunately, Windows platforms are not yet supported by the underlying
+    library.
+
+.. module:: ctc
```
Here too.

notoraptor commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,240 @@
+    Find the directory that contains libwarpctc.so
+    """
+    if config.ctc.root != '':
+        lib_found = False
```
Remove this line (lib_found is used only later, not here).

joaovictortr replied:
Correct, I've removed the variable from that scope.

notoraptor commented on doc/library/tensor/nnet/ctc.txt:
```diff
@@ -17,7 +17,7 @@
     library.
 
-.. module:: ctc
+.. module:: theano.ctc.root
```
It is `theano.tensor.nnet.ctc` here !

joaovictortr replied:
Fixed, I was editing while also editing the description of the PR, I apologize for the mistake.

notoraptor commented:
Hello @joaovictortr see here : https://github.com/Theano/Theano/pull/5949#discussion_r128105369

You should write `theano.tensor.nnet.ctc` at this point, not `theano.ctc.root` !

notoraptor commented:
@joaovictortr Could you rebase this PR ? There is a conflict to resolve.

abergeron commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,174 @@
+    def L_op(self, inputs, outputs, output_grads):
+        # Gradients computed by Op
+        gradients = outputs[1]
+        assert gradients is not None
```
This still doesn't check the length of outputs.

It is acceptable to refuse to provide a gradient it compute_grad is set to False, but check it.

joaovictortr replied:
I've added the checks on both the length and the compute_grad flag on both implementations.

abergeron commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,239 @@
+        ctc_available.msg = 'Windows platforms are currently not supported ',
+        'by underlying CTC library (warp-ctc).'
+        return False
+    elif not ctc_present():
```
ctc_available() should check both that it's present and it compiles.

joaovictortr replied:
In fact, the ``ctc_available`` function calls the ``_ctc_check_compile`` function, which tries to compile a sample code using the CTC library.

joaovictortr replied:
Indeed, do you think it would be best to merge ``ctc_available`` with ``ctc_present``, and provide only ``ctc_available``, for example?

abergeron commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,239 @@
+        return header_dirs
+
+    def c_compile_args(self):
+        return gof.OpenMPOp.c_compile_args(self)
```
Remove this function and it will call the right implementation.

joaovictortr replied:
Done.

abergeron commented on theano/tensor/nnet/ctc.py:
```diff
@@ -0,0 +1,239 @@
+
+    def L_op(self, inputs, outputs, output_grads):
+        gradients = outputs[1]
+        assert gradients is not None
```
Check length here too.

joaovictortr replied:
Done.

joaovictortr commented on theano/configdefaults.py:
```diff
@@ -1772,6 +1772,12 @@ def default_compiledir():
+    'Directory which contains the root of Baidu CTC library. It is assumed \
+    that the compiled library is either inside the build, lib or lib64 \
+    subdirectory, and the header inside the include directory.',
+    StrParam('', allow_override=False))
```
Done.

slefrancois commented:
jenkins test this

notoraptor commented:
Jenkins has successfully run CTC tests:

- http://earlgrey.iro.umontreal.ca:8080/job/Theano_PR/441/testReport/theano.gpuarray.tests.test_ctc/TestCTC/
- http://earlgrey.iro.umontreal.ca:8080/job/Theano_PR/441/testReport/theano.tensor.nnet.tests.test_ctc/TestCTC/

botev commented:
@joaovictortr I just want to say big thanks from someone who will use this! Please do post and a link to any of the examples you have run with this to show it to other people.

joaovictortr commented:
@botev glad to heart it! I hope it is going to be useful for you and other users of Theano! I am having problems finding a dataset with which I can build a real example in Lasagne, do you have any references? I've stumbled upon TIMIT, but it requires licensing.

botev commented:
@joaovictortr I particularly use Libri Speech - http://www.openslr.org/12/ it's free and quite bigger than TIMIT. It is not nessacarily realistic as the speech is clean without noise, but it's about 1000h in total. I don't have a model atm to pass as I'm trying to make a WaveNet Layer for it in Lasagne and am not yet done.

joaovictortr commented:
@notoraptor there was a conflict in ``theano/gpuarray/opt.py``, I've merged the master branch into this branch, but there seems to be a problem with the ``InplaceGpuDimShuffle`` op and the tests fail. Should I try a rebase from the master branch?

joaovictortr commented:
@botev Thanks for the data set! I'll try to build a simpler model in Lasagne.

notoraptor commented:
Hello @joaovictortr ! Always rebase, never add "merge" commits, because at the end we will merge into master branch a branch that contains "merge" commits from that same master branch, that is not really suitable !

Also, remove any "merge" commit from that branch (NB: it may be required by git itself while rebasing).

joaovictortr commented:
@notoraptor rebase done, Git removed the merge commit by itself.

botev commented:
Could we get this merged in soon :pray:

abergeron commented on theano/gpuarray/ctc.py:
```diff
@@ -0,0 +1,175 @@
+
+# Disable gradient computation if not needed
+@register_canonicalize
+@register_stabilize
```
This is not a stabilization optimisation.

abergeron commented:
jenkins test this please

notoraptor commented:
@joaovictortr there are flake8 errors.
```
/home/jenkins/workspace/Theano_PR/theano/tensor/nnet/ctc.py:10:1: F401 'theano.tensor.opt.register_stabilize' imported but unused
/home/jenkins/workspace/Theano_PR/theano/gpuarray/ctc.py:13:1: F401 'theano.tensor.opt.register_stabilize' imported but unused
```

joaovictortr commented:
@notoraptor for some reason the editor did not warn me, they should be fixed now, sorry.

nouiz commented:
Merging as @abergeron and @notoraptor are good with it.
